pkg/ride/generate/internal/constructors_generation.go

```diff
@@ -0,0 +1,290 @@
+package internal
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/wavesplatform/gowaves/pkg/ride/ast"
+)
+
+func constructorName(act actionsObject) string {
+	return strings.ToLower(string(act.StructName[0])) + act.StructName[1:] + "Constructor"
+}
+
+func argVarName(act actionField) string {
+	return act.Name
+}
+
+func extractConstructorArguments(args []actionField) ([]actionField, error) {
+	arguments := []actionField{}
+	seenOrders := map[int]bool{}
+
+	for _, field := range args {
+		if field.ConstructorOrder == -1 {
+			continue
+		}
+		if seen := seenOrders[field.ConstructorOrder]; seen {
+			return nil, errors.Errorf("Duplicate constructor_order: %d", field.ConstructorOrder)
+		}
+		seenOrders[field.ConstructorOrder] = true
+		arguments = append(arguments, field)
+	}
+
+	for i := 0; i < len(seenOrders); i++ {
+		if !seenOrders[i] {
+			return nil, errors.Errorf("constructor_order %d is missing", i)
+		}
+	}
+
+	sort.Slice(arguments, func(i, j int) bool {
+		return arguments[i].ConstructorOrder < arguments[j].ConstructorOrder
+	})
+
+	return arguments, nil
+}
+
+func checkListElementsTypes(cd *Coder, constructorName string, topListVarName string, info *listTypeInfo) {
+	var helper func(listVarName string, info *listTypeInfo)
+
+	helper = func(listVarName string, info *listTypeInfo) {
+		cd.Line("for _, elem := range %s {", listVarName)
+		cd.Line("switch t := elem.(type) {")
+
+		onelineTypes := make([]string, 0, len(info.ElementTypes()))
+		for _, tInfo := range info.ElementTypes() {
+			switch t := tInfo.(type) {
+			case *listTypeInfo:
+				cd.Line("case %s: ", t.String())
+				helper("t", t)
+			default:
+				onelineTypes = append(onelineTypes, t.String())
+			}
+		}
+
+		cd.Line("case %s:", strings.Join(onelineTypes, ","))
+		cd.Line("default:")
+		cd.Line("return nil, errors.Errorf(\"%s: unexpected type '%%s' in %s list\", t.instanceOf())", constructorName, topListVarName)
+		cd.Line("}")
+		cd.Line("}")
+	}
+	helper(topListVarName, info)
+}
+
+type constructorStructInfo struct {
+	rideName   string
+	goName     string
+	argsNumber int
+}
+
+type versionInfo struct {
+	version        ast.LibraryVersion
+	newStructs     []constructorStructInfo // new structs or modified structs
+	removedStructs map[string]bool         // structs removed in this version
+}
+
+func newVersionInfo(version ast.LibraryVersion) *versionInfo {
+	return &versionInfo{
+		version:        version,
+		newStructs:     make([]constructorStructInfo, 0),
+		removedStructs: make(map[string]bool),
+	}
+}
+
+type versionInfos map[ast.LibraryVersion]*versionInfo
```

Comment:
Why values are pointers to `versionInfo`? Only because of return type of `newVersionInfo` function? Can't find places where it can be `nil`.